hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java

```diff
@@ -410,61 +416,183 @@ public boolean rename(final Path src, final Path dst) throws IOException {
     Path qualifiedSrcPath = makeQualified(src);
     Path qualifiedDstPath = makeQualified(dst);
 
+    Path nestedDstParent = dst.getParent();
+    if (getAbfsStore().getAbfsConfiguration().getPrefixMode()
+        == PrefixMode.BLOB) {
+      /*
+       * For blob endpoint with non-HNS account, client has to ensure that destination
+       * is not a sub-directory of source.
+       */
+      LOG.debug("Check if the destination is subDirectory");
+      while (nestedDstParent != null) {
+        if (makeQualified(nestedDstParent).equals(qualifiedSrcPath)) {
+          //testRenameChildDirForbidden.
+          LOG.info("Rename src: {} dst: {} failed as dst is subDir of src",
+              qualifiedSrcPath, qualifiedDstPath);
+          return false;
+        }
+        nestedDstParent = nestedDstParent.getParent();
+      }
+    }
+
+
     TracingContext tracingContext = new TracingContext(clientCorrelationId,
         fileSystemId, FSOperationType.RENAME, true, tracingHeaderFormat,
         listener);
     // rename under same folder;
-    if(makeQualified(parentFolder).equals(qualifiedDstPath)) {
+    if (makeQualified(parentFolder).equals(qualifiedDstPath)) {
       return tryGetFileStatus(qualifiedSrcPath, tracingContext) != null;
     }
 
-    FileStatus dstFileStatus = null;
     if (qualifiedSrcPath.equals(qualifiedDstPath)) {
       // rename to itself
       // - if it doesn't exist, return false
       // - if it is file, return true
       // - if it is dir, return false.
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
-      if (dstFileStatus == null) {
+      final AtomicBoolean isDstDirectory = new AtomicBoolean();
+      final AtomicBoolean isDstExists = new AtomicBoolean();
+
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
+      if (!isDstExists.get()) {
         return false;
       }
-      return dstFileStatus.isDirectory() ? false : true;
+      return isDstDirectory.get() ? false : true;
     }
 
+    final AtomicBoolean isDstDirectory = new AtomicBoolean();
+    final AtomicBoolean isDstExists = new AtomicBoolean();
+
     // Non-HNS account need to check dst status on driver side.
-    if (!abfsStore.getIsNamespaceEnabled(tracingContext) && dstFileStatus == null) {
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
+    if (!abfsStore.getIsNamespaceEnabled(tracingContext)) {
```

Comment:
can't understand the reference. Kindly share the code pointer please. Thank you.

Reply:
Since, the files have changed a bit, sharing the link to the commit being reviewed: https://github.com/ABFSDriver/AbfsHadoop/blob/2e9a85cda4a0294ea4e9398a5ed6a84ed03649da/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java#L420.

Reply:
PrefixMode checks if its FNS + Blob. The client has to make qualifiedDst in case of FNS(let it be DFS or BLOB). Ex: rename(/srcDir, /dstDir) -> the qualifiedDst will be `/dstDir/srcDir`.